### ISSUE TITLE
Add builder documentation

### ISSUE THREAD

**Issue by Und3rf10w** [Und3rf10w/external_c2_framework#19]
While #17 introduced a lot of new and highly desired functionality, this unfortunately, made almost all of the existing documentation obsolete.

New documentation must be updated to match the current application.

**Comment by Und3rf10w** [comment]
Most of the critical documentation I want to add is below, but it isn't through or complete enough to actually be commited yet.

## Builder
The builder reads in a config file, and uses configured options to generate a build by replacing `markers` within `skeletons`.

The builder can be used with `build_files.py`. A sample builder configuration is provided as `sample_builder_config.config.sample`.

## Skeletons
`skeletons` are the different "skeletons" of code that the `builder` will dynamically populate to generate a completely usable build. `skeletons` contain `markers` that will be replaced with usable values by the `builder`. There are three different 'types' of skeletons:

1. Frameworks
2. Transports
3. Encoders

### Skeleton Markers
A marker can be placed inside any file in a `skeleton`, which will be replaced with a value specified in the builder config. In best practice, **markers should never be used to directly write variables, and should only ever be used to set values**. If a marker's value has to be reused, one should opt to store the value in a variable and reference it that way, instead of reusing the same marker.

The marker format is: ` ```[var:::identifier_for_the_marker]``` `

Strings will be written into a skeleton directly wrapped in single quotes, and numbers will be written as is.

**In the event a string in the config is wrapped in double quotes**, the string will be written directly to the file wrapped in double quotes instead, and the wrapping single quotes will be stripped.

This relationship can demonstrated as:

```python
#################
# Skeleton Code #
#################

# Skeleton contains the following line of code:
foo = ```[var:::bar]```

##############
# End Result #
##############

# Stored in config as:
#   foo = bar
# Written as:
foo = 'bar'

# Stored in config as:
#   foo = "bar"
# Written as:
foo = "bar"

# Stored in config as:
#   foo = 2
# Written as:
foo = 2

# Stored in config as:
#  foo = "2"
# Written as:
foo = "2"
```

### Frameworks
Frameworks are the base application that determines what data is being used by the `transport` and `encoder`, and how that data is used. What a specific `framework` actually does doesn't really matter, so long as logic exists to import and use the `encoder` and `transport`. 

There are things to consider when building a `framework`:
* The `framework` is responsible for ensuring that the `encoder` is made available to the `transport` to be used.
* If the `framework` uses a client-server relationship, they should be appropriately organized as such.
* A `framework` skeleton will be processed by the builder, iterating through every file in it, so if a certain argument needs to be configurable, it can be easily used.

### Encoders
Encoders define how raw data is modified before being sent through, or after being received from the `transport` channel. This could be anything as simple as base64ing the data, to something more complicated such as embedding the data into a file.

Care should be taken to design encoders to handle data neutrally, so that they can be reused across different `frameworks`. It is critical to maintain the integrity of the data, and modify it in a fashion that can restore the data to its raw format that a `framework` requires, whatever that may be.

### Transports
Transports define how data is transmitted through a communication channel. Transports shouldn't directly modify data in any fashion, as that should be handled entirely by the `encoder`. For example, if a `transport` has to make a specific HTTP request, it can EMBED the data wherever is appropriate in the web request, but the process of modifying the data into a form that can be used for the request should be handled by the `encoder`.

Originally, transports were designed to receive already encoded data, but this caused issues when trying to encode data using methods that would require storing a large amount of data in memory (e.g. embedding data in multiple high resolution images). Now, transports should expect the `framework` to make the `encoder` available to so that the `transport` may utilize the functions to modify the data as appropriate. In other words, **the `transport` is responsible for calling the `encoder`, and the `framework` is responsible for ensuring that the `encoder` is available to use for the `transport`.**

Transports should expect to be able to call `encoder` functions as `encoder.$FUNCTION_NAME()`, and encoder variables as `encoder.$VARIABLE_NAME`.

**Comment by Und3rf10w** [comment]
Fixed by a77d863
